cli/bpbuild/build.go

```diff
@@ -0,0 +1,151 @@
+	appendClientFilteredBuilds := func(builds []*cloudbuild.Build) {
+		for _, b := range builds {
+			appendBuild := true
+			for _, cFilter := range cFilters {
+				// skip if any client side filter evaluates to false
+				if !cFilter(b) {
+					appendBuild = false
+					break
+				}
+			}
+			if appendBuild {
+				cbBuilds = append(cbBuilds, b)
+			}
+		}
+	}
+
+	if len(c.Builds) < 1 {
+		return nil, fmt.Errorf("no builds found with filter %s in project %s", filter, projectID)
+	}
+	appendClientFilteredBuilds(c.Builds)
+
+	// pagination
+	for {
+		c, err = cloudbuildService.Projects.Builds.List(projectID).Filter(filter).PageToken(c.NextPageToken).Do()
+		if err != nil {
+			return nil, fmt.Errorf("error retriving next page with token %s: %v", c.NextPageToken, err)
+		}
+		appendClientFilteredBuilds(c.Builds)
+		if c.NextPageToken == "" {
+			break
+		}
+	}
+	return cbBuilds, nil
+}
+
+// clientside filter functions
+type clientBuildFilter func(*cloudbuild.Build) bool
+
+// filterRealBuilds filters out builds not triggered from source repos (i.e by automation).
+func filterRealBuilds(b *cloudbuild.Build) bool {
+	for _, subs := range []string{"COMMIT_SHA", "REPO_NAME", "TRIGGER_NAME"} {
+		_, substExists := b.Substitutions[subs]
+		if !substExists {
+			return false
+		}
+	}
+	return true
+}
+
+// filterGHRepoBuilds filters builds from a particular repo name.
+// TODO:(bharathkkb): We should ideally be using a sever side filter for this https://cloud.google.com/build/docs/view-build-results#filtering_build_results_using_queries
+// but I was not able to figure out expected format for GH URLs.
+func filterGHRepoBuilds(repo string) clientBuildFilter {
+	return func(b *cloudbuild.Build) bool {
+		name, exists := b.Substitutions["REPO_NAME"]
+		if !exists {
+			return false
+		}
+		return name == repo
+	}
+}
+
+// successBuildsBtwFilterExpr returns a CEL expression as string
+// for finding all successful builds between start and end time.
+func successBuildsBtwFilterExpr(start, end time.Time) string {
+	return fmt.Sprintf(
+		"create_time>=\"%s\" AND create_time<\"%s\" AND status=\"%s\"",
+		start.Format(time.RFC3339),
+		end.Format(time.RFC3339),
+		successStatus)
+}
+
+// getBuildFromFile unmarshalls a CloudBuild file at path.
+func getBuildFromFile(path string) (*cloudbuild.Build, error) {
+	content, err := os.ReadFile(path)
+	if err != nil {
+		return nil, err
+	}
+	var b cloudbuild.Build
+	err = yaml.Unmarshal(content, &b)
+	if err != nil {
+		return nil, err
+	}
+	return &b, nil
+}
+
+// getBuildStepIDs retrieves a slice of build step IDs in a build.
+func getBuildStepIDs(b *cloudbuild.Build) []string {
+	steps := []string{}
+	for _, bs := range b.Steps {
+		steps = append(steps, bs.Id)
+	}
+	return steps
+}
+
+// findBuildStageDurations computes duration for a given build stage across a slice of builds
+// iff stage is successful.
+func findBuildStageDurations(stepId string, builds []*cloudbuild.Build) ([]time.Duration, error) {
```

Comment:
Hmm my thought was
- cloudbuild.Build is from 3p library so we will need to wrap if we need to add a method
- we deal with n builds so I it does not benefit from access to info about a single build

WDYT?

Reply:
Agreed that it'll be a bit more work to write a wrapper but it'll make thing more clearer and the fact that you'll get rid of the embedded loops is nice too. I'll let you make the decision if you're willing to spend more time of keep it as is. Feel free to resolve this if you have no further comments.

Reply:
In this case, I am leaning towards keeping it a function because we will still need to do a nested loop (as we are trying to find 1 build step per n build steps across n builds) and seems to add boilerplate. Let's re evaluate when we add more functionality.